src/librustc_lint/methods.rs

```diff
@@ -0,0 +1,102 @@
+use crate::LateContext;
+use crate::LateLintPass;
+use crate::LintContext;
+use rustc_hir::{Expr, ExprKind};
+use rustc_middle::ty;
+use rustc_span::{
+    symbol::{sym, Symbol, SymbolStr},
+    ExpnKind, Span,
+};
+
+declare_lint! {
+    pub TEMPORARY_CSTRING_AS_PTR,
+    Deny,
```

Comment:
We'll need a crater run to check the fallout from this (but it should still be deny by default unless we break some foundational crate, IMO).

Reply:
dependencies get their lints capped anyway, so it will only break building that crate directly.

Reply:
I feel that it shouldn't be deny by default as long as it considers `func(CString::new("...").unwrap().as_ptr());` to be erroneous (since in typical situations this would be a false positive).

Reply:
+1 for making this Warn, and not Deny (though I don't know if we have other similar deny-by-default lints). 

I think we also should add a regression test for `func(CString::new("...").unwrap().as_ptr());`. I wouldn't personally call it as  a false-positive though: it's not UB, but still sufficiently obscure style worth linting against.

Reply:
I agree on the Warn by default, since the lint covers cases that aren't *incorrect*. That's reflected in the PR now.